docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,225 @@
+
+#### As an operator, I want to be able to have pods run quicker when scheduled
+
+As a cluster operator I want to bring up new nodes that already have the container images of the applications it will run pre-cached so that pods start running quicker when scheduled.
+
+#### Credits
+
+A big thank you to Jack Francis, Michael Sinz, and Amr Hanafi for their work on the Kamino project that used a similar Prototype Pattern with AKSEngine which inspired this feature request: jackfrancis/kamino: Kubernetes node clone factory (github.com)
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+None currently known to be needed
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+##### NFR1
+The solution should increase the performance of horizontal scaling over time as people use this feature
+
+##### NFR2
+The solution should cause no increased risks with security or reliability
+
+### Implementation Details/Notes/Constraints
+
+The plan is to create a Controller which implements the feature with the Node Prototype Pattern as desired.
+
+These controller additions can be added to azuremachine_controller.go because that is where AzureMachines are being controlled and azuremachinepool_controller.go since that is where the AzureMachinePools are being controlled.
+
+An operator will be able to decide to turn the feature on or off with a flag on creation of a cluster, and then can alter the YAML file for AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be.
+
+Every time that amount of time has passed, it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image.
+
+	As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachineTemplates.
+
+In terms of how a healthy node would be selected, there should be an annotation on each node as to when they’ve last been patched. An ideal node would be one which has been patched since the last prototype went into service and is running and healthy. Whichever node has been steadily healthy for the longest amount of time should be chosen as it’s the most overall stable. As the prototype is always from a successfully healthy and working node the image is always known to be working before being chosen for replication.
```

Comment:
Made this more clear in that we simply want to know how long it's been since the last patch(s) for nodes who have that patch(s), a reset to the clock should naively reset how long it has been stable since differentiating between why it reset is out of scope/much nastier to do and not necessary for a naïve implementation (assuming I understood this comment correctly)